src/metabase/lib/field.cljc

```diff
@@ -529,3 +530,137 @@
+  (let [stage (lib.util/query-stage query stage-number)]
+    (case (:lib/source column)
+      :source/table-defaults      (if (contains? stage :fields)
+                                    (append-field query stage-number column)
+                                    ;; If :fields key is not set on the query, it's implicitly :all, so nothing to add.
+                                    query)
+      :source/joins               (add-field-to-join query stage-number column)
+      :source/implicitly-joinable (append-field query stage-number column)
+      ;; Default case - for columns from a card, from a native query, from a custom expression - these are always
+      ;; returned and cannot be selected off and on.
+      query)))
+
+(defn- remove-top-field
+  "This is called only for fields that plausibly need removing. If the stage has no `:fields`, this will populate it.
+  It shouldn't happen that we can't find the target field, but if that does happen, this will return the original query
+  unchanged. (In particular, if `:fields` did not exist before it will still be omitted.)"
+  [query stage-number column]
+  (let [populated  (if (:fields (lib.util/query-stage query stage-number))
+                     query
+                     (populate-fields-for-stage query stage-number))
+        old-fields (:fields (lib.util/query-stage populated stage-number))
+        column-ref (lib.ref/ref column)
+        index      (first (keep-indexed (fn [index field]
+                                          (when (lib.equality/find-closest-matching-ref query column-ref [field])
+                                            index))
+                                        old-fields))
+        fields'    (when index
+                     (let [[pre [_ & post]] (split-at index old-fields)]
+                       (vec (concat pre post))))]
+    ;; It's still possible
+    (if index
+      (lib.util/update-query-stage populated stage-number assoc :fields fields')
+      query)))
+
+(defn- remove-field-from-join [query stage-number column]
+  (let [field-ref    (lib.ref/ref column)
+        join         (lib.join/resolve-join query stage-number (::lib.join/join-alias column))
+        join-fields  (lib.join/join-fields join)]
+    (if (or (nil? join-fields)
+            (= join-fields :none))
+      ;; Nothing to do if there's already no join fields.
+      query
+      (let [join-fields (if (= join-fields :all)
+                          (map lib.ref/ref (lib.metadata.calculation/returned-columns query stage-number join))
+                          join-fields)
+            removed     (remove #(lib.equality/find-closest-matching-ref query field-ref [%]) join-fields)]
+        (if (= (count join-fields) (count removed))
+          ;; Return the query unchanged if we didn't find anything.
+          query
+          (lib.remove-replace/replace-join query stage-number join (lib.join/with-join-fields join removed)))))))
```

Comment:
Another `cond->`. (In my reading that would correspond to the "Return the query unchanged" part of the comment.)